### PR TITLE
Issue #11 Update test script functions

### PR DESCRIPTION
**Pull request by araza7865**
Addresses Issue #11 and #26.  Follows from https://github.com/LBNL-ETA/guideline36_conformance_test/pull/31.

This PR implements a new OOP-based framework for handling string-defined operations in test expressions, addressing Issue #11.

### Summary of changes

1. **Introduced state-based operation base class**  
   - Added `StateOperation` as a common base for time-dependent operations that maintain state and act on a test variable.
   - Encapsulates:
     - Raw expression string
     - Reference to the `test` object
     - Target `variable`
     - Parsed `params` and `computed_value`

2. **Implement `Ramp` and `Periodic` as stateful operations**  
   - `Ramp`:
     - Parses `RAMP(start; end; duration; period)` style expressions.
     - Computes:
       - `ramp_start`, `ramp_end`, `duration`, `ramp_rate`, `ramp_period`, `ramp_step`
     - On `set_value(seconds_since_start)`, updates the target variable on period boundaries, ramping up or down until the end value is reached.
     - Tracks instances (`Ramp.instances`) with `destroy()` / `destroy_all()` helpers for lifecycle management.
   - `Periodic`:
     - Parses `PERIODIC(expression; period)` style expressions.
     - Stores `periodic_expression` and `period`, and reevaluates the expression at the configured period.
     - Uses the same instance tracking and lifecycle helpers as `Ramp`.

3. **Introduce stateless operation hierarchy for pure expressions**  
   - Added `StateLessOperation` as a base for operations that:
     - Take a raw string and `test` object.
     - Parse parameters once and compute a `computed_value` without internal state across timesteps.
   - Added `TwoTermOperation` to encapsulate the common pattern:
     - Parse `OP(arg1; arg2)`
     - Evaluate both arguments via `test.evaluate_expression`
     - Delegate to `_apply(a, b)` for the actual operation.
   - Implemented concrete two-term operations:
     - `Add` – handles `ADD(a; b)` → `a + b`
     - `Sub` – handles `SUB(a; b)` → `a - b`
     - `Mul` – handles `MUL(a; b)` → `a * b`

4. **Add interpolation operation**  
   - Implemented `InterpolateOperation` to handle:
     - `INTERPOLATE(x; x0; x1; y0; y1[; min_out[; max_out]])`
   - Behavior:
     - Linearly interpolates between `(x0, y0)` and `(x1, y1)` at `x`.
     - Optionally clamps the result between `min_out` and/or `max_out` when provided.
   - Integration:
     - Uses `test.evaluate_expression` for all arguments, consistent with other operations.
     - Stores parsed values in `self.params` and computes the final `computed_value` via `_apply()`.

5. **Handling "=ANY"**
 - Corrected the previous check on the presence of the string "Any" in the test script to the case-corrected version "ANY", as used in the most recent version of the test script.
---

### Notes and follow-ups

- The current implementation focuses on structure and correctness of the expression handling and time-based behaviors.
- **Unit testing / performance**:
  - There is currently no dedicated unit test suite for these operations.
  - A follow-up task is needed to:
    - Add unit tests for all operations (`Ramp`, `Periodic`, `Add`, `Sub`, `Mul`, `InterpolateOperation`) covering both parsing and evaluation.
    - Include tests for edge cases (e.g., zero/negative periods, identical `x0`/`x1`, boundary conditions for clamping, float rounding behavior).
    - Handle percentages

### CODE REVIEW

**Comment by araza7865** [comment]
Currently fails Test A3 with the message "For variable con.yDam [or VAVDmprCmd], actual value = 0.101335 not = expected value = 0.000000"

**Comment by dhblum** [comment]
@araza7865 FYI with PR https://github.com/LBNL-ETA/guideline36_conformance_test/pull/20 closed I merged the latest ``develop`` branch into this development branch, resolved conflicts, and pushed the result.  So feel free to pull the latest and continue working from there.

**Comment by dhblum** [comment]
@araza7865 Actually, sorry, there's some work to do to get the dummy test script etc. working with the updates from the latest devleop.  Still working through some things.

**Comment by dhblum** [comment]
@araza7865 Ok, sorry for some excessive messaging, but this should be good to run with the appropriate dummy test script, point map, and modelica files, which I'll email separately for now.  Note that I've created a new conformance_test called "foo", replacing the name "dummy."

File: src/Test.py [hunk 1]
```diff
@@ -179,9 +175,24 @@ def start_test(self, to_csv=False, name=None):
             print("Successfully set input values=================================")
             print()
 
-            step_start_time = self.controller.get_current_time()
+            step_start_time = self.controller.get_current_time()            
+            if self.current_step == 1 and self.controller.get_type() == 'simulation':
+                self.controller._initialize_sim()
```
Comment (dhblum) [inline src/Test.py]: All this from lines 179-186 that uses if-else to handle simulation-based initialization and getting current time was erased as part of Issue #10 in [PR #20](https://github.com/LBNL-ETA/guideline36_conformance_test/pull/20/changes#diff-81684c38727697def37693cc6c4204401e785792d380e67708e54d58517cd9e4L137).  Why is it being put back?  Check that it can be removed since we want to abstract device types away from Test.py.

File: src/Test.py [hunk 2]
```diff
@@ -391,10 +368,24 @@ def test_conditions(self, condition, st, sleep_interval=None, verbose=False, to_
...
+        for obj in Ramp.instances:
+            if obj.params['ramp_step']:                        
+                obj.set_value(seconds_since_start)                    
+
+        for obj in Periodic.instances:
+            # import pdb; pdb.set_trace()
+            if obj.params['periodic_step']:
+                obj.set_value(seconds_since_start)
+    
+        self.controller.final_step_wait(flex_duration = 0.0000001)
```
Comment (dhblum) [inline src/Test.py]: This won't work for bacnet-based devices since ``final_step_wait()`` is not part of device base class.  I'm realizing the ``wait()`` function has a parameter ``duration``, which could probably be used instead of creating a whole new function ``final_step_wait()``, maybe with slight modifications.  This would allow for better device abstraction.  Waiting a small amount of time for bacnet devices, like this implementation for simulation-based devices intends, should probably be fine too.

**Review by dhblum — CHANGES_REQUESTED** [review]
Leaving an initial review for now to start a to-do list.  Not yet totally complete.

File: src/Test.py (hunk 1 above)
Comment (dhblum) [reply src/Test.py]: Addressed in https://github.com/LBNL-ETA/guideline36_conformance_test/pull/18/commits/32b601a8989cb9557f03c3b6a4b681548e60508a.

File: src/Test.py (hunk 2 above)
Comment (dhblum) [reply src/Test.py]: Addressed in https://github.com/LBNL-ETA/guideline36_conformance_test/pull/18/commits/27dfff8908b4f95019574944a8e7fd30436deac0.

**Review by dhblum — APPROVED** [review]
Made some changes, but good to go.
